docs/spec/lightclient/attacks/evidence-handling.md

```diff
@@ -0,0 +1,250 @@
+            attackType = EquivocationAttack
+        } else {
+           attackType = AmnesiaAttack
+        }                                        
+                 
+        return (LightClientAttackEvidence { conflictingBlock, trustedBlock.Header.Height, attackType }, 
+                trace(lightStore, trace[i-1].Height, trace[i].Height))
+    } 
+    return (nil, nil)       
+}
+```
+
+## Evidence handling 
+
+As part of on chain evidence handling, full nodes identifies misbehaving processes and informs
+the application, so they can be slashed. We now specify for each event type evidence handling logic.
+
+### Lunatic attack evidence handling
+
+```go
+func detectMisbehavingProcesses(lcAttackEvidence LightClientAttackEvidence, bc Blockchain) []ValidatorAddress {
+   assume lcAttackEvidence.Type == LunaticAttack
+   misbehavingValidators = []ValidatorAddress
+
+   conflictingHeight = lcAttackEvidence.ConflictingBlock.Header.Height
+   
+   bondedValidators = bc.GetValidators(lcAttackEvidence.CommonHeight) union 
+                      bc.GetNextValidators(lcAttackEvidence.CommonHeight) union
+                      bc.GetValidators(conflictingHeight)) 
+   
+   return getSigners(lcAttackEvidence.ConflictingBlock.Commit) intersection bondedValidators        
+}
+
+func getSigners(commit Commit) []ValidatorAddress {
+    signers = []ValidatorAddress
+    for (i, commitSig) in commit.Signatures {
+        if commitSig.BlockIDFlag == BlockIDFlagCommit { 
+            signers.append(commitSig.ValidatorAddress)                    
+        }
+    }
+    return signers
+}
+```
+
+### Equivocation attack evidence handling
+
+```go
+func detectMisbehavingProcesses(lcAttackEvidence LightClientAttackEvidence, bc Blockchain) []ValidatorAddress {
+   assume lcAttackEvidence.Type == EquivocationAttack
+   
+   conflictingBlock = lcAttackEvidence.ConflictingBlock 
+   trusted = bc[conflictingBlock.Header.Height+1].LastCommit 
+   conflicting = conflictingBlock.Commit
+   
+   return getSigners(trusted) intersection getSigners(conflicting)      
+}
+```
+
+### Amnesia attack evidence handling
+
+Detecting faulty processes in case of the amnesia attack is more complex and cannot be inferred 
+purely based on attack evidence data. In this case, in order to detect misbehaving processes we need
+access to votes processes sent during the conflicting height. Therefore, amnesia handling assumes that
+validators persist all votes received and sent during multi-round heights (as amnesia attack 
+is only possible in heights that executes over multiple rounds, i.e., commit round > 0).  
+
+To simplify description of the algorithm we assume existence of the trusted oracle called monitor that will 
+drive the algorithm and output faulty processes at the end. Monitor can be implemented in a
+distributed setting as on-chain module. The algorithm works as follows:
+    1) Monitor sends votesets request to validators of the conflicting height. Validators
```

Comment:
do we really need to make requests? is it possible to, when amnesia attack evidence is committed, to say that suspected nodes must gossip their votesets. If they don't, we punish them.

Reply:
Sure. Making requests is more conceptual step here; there are probably more efficient version to implement it.